### PR TITLE
Merged PR#30 and added a camera tab in configuration

### PR DESCRIPTION
**Pull request by codename5**
Merged PR#30 into current main branch. This allows the user to add an image file to any of the 3 cameras in addition to the default blank image. Only PNG is supported as of this time.

Added a camera tab in configuration so the user can change the image at will (check below).

image

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3451)

### CODE REVIEW

**Comment by chris062689** [comment]
Is the outer left camera and outer right supposed to combine for the 3D effect?
We might want to put a label somewhere to designate their usage.

**Comment by codename5** [comment]
I haven't tested it for 3D. I've used it for scanning qr codes mostly. I've tested it on Pokemon Alpha Sapphire and Pokemon Moon. Pokemon AS uses outer left camera while Pokemon M uses outer right camera.

File: src/citra_qt/configuration/configure.ui [hunk 1]
```diff
@@ -6,7 +6,7 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>461</width>
+    <width>580</width>
```
Comment (wwylele) [inline src/citra_qt/configuration/configure.ui]: Is this change intentional?

File: src/citra_qt/configuration/configure_camera_individual.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    ui->setupUi(this);
+}
+
+ConfigureCameraIndividual::~ConfigureCameraIndividual() {}
+
+void ConfigureCameraIndividual::setConfiguration() {
+    int camera_mode_index = 0;
+    for (int index = 0; index < ui->camera_mode_combobox->count(); index++) {
+        if (ui->camera_mode_combobox->itemText(index).toStdString() ==
+            Settings::values.camera_name[m_CameraId]) {
```
Comment (wwylele) [inline src/citra_qt/configuration/configure_camera_individual.cpp]: This would not work if the UI text in the combo box is translated. The `camera_name` is only for internal use and shouldn't be bound to UI text. You can give items in the combobox some internal user data (iirc Qt provides this kind of API) and do name matching from there.

File: src/citra_qt/configuration/configure_camera_individual.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    ui->camera_mode_combobox->setCurrentIndex(camera_mode_index);
+    UpdateCameraMode(camera_mode_index);
+
+    ui->image_path->setText(Settings::values.camera_config[m_CameraId].c_str());
+}
+
+void ConfigureCameraIndividual::applyConfiguration() {
+    Settings::values.camera_name[m_CameraId] =
+        ui->camera_mode_combobox->itemText(ui->camera_mode_combobox->currentIndex()).toStdString();
```
Comment (wwylele) [inline src/citra_qt/configuration/configure_camera_individual.cpp]: ditto about the UI text vs internal text problem.

**Comment by wwylele** [comment]
For anyone wondering, this is based on the branch in my repo (https://github.com/wwylele/citra/pull/30, hence "PR#30"). However, I already forget many stuff there so I can't take the responsibility to address issues in the code.

----

> Is the outer left camera and outer right supposed to combine for the 3D effect?
Yes. However, if the game only wants a 2D image, it can choose receiving image from either camera.

**Comment by jroweboy** [comment]
I'll give a thorough code review later, but i wanted to first say that I think the UI could use some work. 

First off, theres no reason it should get its own tab. I would expect something about cameras to go in the system tab as its part of the 3ds system emulation. 

Second, I think a better way to do the UI would be to have two groups, Front (Face) Cam and Back Cam. The Front cam options would look like the custom camera widget you built there, while the Back Cam would have a selector for "2D" or "3D" mode. 

In 2D mode, it will look like the front camera widget, where it shows only one camera selection widget, and on the backend of the code, you pass this camera device to both left and right camera. In 3D mode, you get something like the design you posted, with both left and right camera being configurable. I can't really see 3d mode being all that useful, as most games don't use the full 3d camera, which is why i think we should take the opportunity to make it easier to configure for most people. This way they configure just one camera, and they get their qr code scanned, and don't have to figure out if the game checks the left or the right camera. But also, we have the ability to configure each camera separately if the game requires it.

Let me know if you have any questions, and also thanks for the PR ❤️

**Comment by wwylele** [comment]
I do think camera deserves its own tab, as we might going to add more sophisticated configuration, such as device selection for a real camera and calibration etc. They are going to fill up the entire window.

File: src/citra_qt/configuration/configure_camera_individual.cpp (hunk 2 above)
Comment (codename5) [reply src/citra_qt/configuration/configure_camera_individual.cpp]: Fixed. Made the setting save into data property of combobox.

File: src/citra_qt/configuration/configure_camera_individual.cpp (hunk 3 above)
Comment (codename5) [reply src/citra_qt/configuration/configure_camera_individual.cpp]: Ditto fixed :)

File: src/citra_qt/configuration/configure.ui (hunk 1 above)
Comment (codename5) [reply src/citra_qt/configuration/configure.ui]: 8 tabs don't fit on 461 width. It needs at least 565 width. I personally find it annoying to move to later tabs or resize the menu. It doesn't save too when reopened.
It is up to you but IMO someone needs to redesign the tabs.

**Comment by codename5** [comment]
QImage needs the qt image formats plugin to support other formats. I'll see how I can add it.

For the configuration tabs, the whole window definitely needs a redesign in the future when more settings are available. I'll leave it to you guys.

I think Camera should be a separate from configuration. I'd place it under emulation. It is kinda like cheats. You'll open it a lot when you are using it especially if you want to scan multiple qr images or you want to close the camera (when it is added). I'd like to hear your suggestion on this so I can decide whether or not to move the tab.

For 3D, I am not an expert but I think the correct way to design it would be to supply one image file and Citra will split it for you and feed it to each camera.  I am personally not interested in doing 3D, so I can't do much beyond the UI. I hope you can understand.

For changes in UI, below is a mock design made in qtcreator. I'd add Front Camera below it or if we're going to use a separate window, we can split them into tabs and add a third tab for shared settings.

image

**Comment by Pikacciu** [comment]
Hello,

Is it possible to add a virtual pad control layout for tactile screen on Citra ?

Will provide a better experience of gaming than with a keyboard

**Comment by BreadFish64** [comment]
@Pikacciu Please avoid commenting in unrelated pull requests, if you have a feature request it might be more welcome in the [community forums](https://community.citra-emu.org/)

**Comment by jroweboy** [comment]
> I'd like to hear your suggestion on this so I can decide whether or not to move the tab.

Keep it in its own new tab for now. I personally dislike this as we have too many tabs as it is, but whatever. It will still look pretty bad on other translations I'm guessing, but no one seems to care :P 

> For 3D, I am not an expert but I think the correct way to design it would be to supply one image file and Citra will split it for you and feed it to each camera. I am personally not interested in doing 3D, so I can't do much beyond the UI. I hope you can understand.

I think my last post was confusing. Let me clarify with pictures. All I was getting at was keeping the same functionality you have here, but simplifying it for the users. Most people don't want to fiddle with "Back Left Camera" and "Back Right Camera" to find which one the game uses to read a single 2d image. So I proposed a "simple" config (called 2d) where a single input source is used for both back cameras, and a slightly more complicated config (called 3d) where you can choose the image for each camera individually. If you can think of a better name than 2D or 3D for the mode option, then be my guest to change it. Heres a mock up of what I pictured for 2D mode (aka a single image is used for both camera as input)

qtcreator_2018-02-21_09-56-55

The basis is that most games don't use both feeds from the back camera, so most users shouldn't care about configuring two cameras as well.  I don't see a reason to display two inputs with one disabled when we can just show one that'll work for most every game out there. 

The 3d mode would be basically what you posted above just with left enabled :) That way if a game DOES read both sources and looks to piece together something in 3d, then they'll be able to work something out.
  


(And before you say "Would people really care if i just left the 'Left Camera' disabled instead of hiding it" lemme tell you from personal experience that having an option disabled in the config like vsync for instance does lead to people complaining about it on the forums)

Lastly, I'm just trying to propose an easier to use system. I'm open to other suggestions, but I very much want to avoid the scenario you described in an earlier message. "Pokemon AS uses outer left camera while Pokemon M uses outer right camera. So you'll need all cameras working even for 2D." <- We won't need this if we run the single image through both cameras, meaning users won't have to play the guessing game to figure out which camera the game uses.

Also all the stuff about a separate window was talking about the hypothetical more complicated config that wwylele brought up. I'm not suggesting in any way that this needs its own window or anything. Something to allow quick access to settings should be designed independently from this PR imo :)

**Comment by adityaruplaha** [comment]
Imo, this does deserve it's own window, as the config is getting pretty cluttered up. But for now,  a tab's alright.

**Comment by jroweboy** [comment]
@codename5 Sorry to bother you, are you still planning on finishing this PR? I'd really like to see this merged 👍

**Comment by adityaruplaha** [comment]
So will I. Please clean it up. I've seen this feature in unofficial builds for years & comments following this often contain really hurtful remarks on official Citra's usefulness.

**Comment by adityaruplaha** [comment]
@codename5 are you still on this?

**Comment by Robz8** [comment]
Hopefully someone picks this up.
Would love to try the AR features (assuming those work) of the Style Savvy games with this.

**Comment by zhaowenlan1779** [comment]
I've been working on this one for nearly a week. I didn't see this PR and did it myself...

I've already implemented a camera with the opencv library, with a better configuration ui (or at least I think so). I think there's no need to have '2d' or '3d' configuration (it will only make things more complicated). I used a QComboBox to switch between different cameras.

To make the configuration page more user-friendly, I added an 'Image Source' selection.

Current status: It works fine for most games like Pokemon SM/USUM. But for some games(like Pokemon Mystery Dungeon GTI) it's very slow. I'm currently trying to imporve this.

image

(The image is too big..)

**Comment by jroweboy** [comment]
> I think there's no need to have '2d' or '3d' configuration (it will only make things more complicated).

Your current design doesn't address the issue that I addressing with that suggestion. 

> Lastly, I'm just trying to propose an easier to use system. I'm open to other suggestions, but I very much want to avoid the scenario you described in an earlier message. "Pokemon AS uses outer left camera while Pokemon M uses outer right camera. So you'll need all cameras working even for 2D." <- We won't need this if we run the single image through both cameras, meaning users won't have to play the guessing game to figure out which camera the game uses.

With your design, the original problem that the users don't know what camera to configure for what game is back. They just have to blindly configure all the cameras instead of having a mode that lets them setup a single image to be shared through all cameras.

As i mentioned, I hate the name 2d and 3d. if you think of a better name, go ahead and change it. But having people tweak each camera just to scan their pokemon qr codes is complicated when we can design it to be easy to config for 90% of the use cases. As most games only use one of the cameras, we should design the interface to be simple for this use case.

**Comment by Zantar** [comment]
Hello.
First I just want to say sorry if this is not the right place to post this. I just got a tip I should post this here so sorry if that is not the case.
So I downloaded a version of Citra with camera support from [https://www.reddit.com/r/Citra/comments/7yulns/build_latest_citra_canary_with_camera_image/](https://www.reddit.com/r/Citra/comments/7yulns/build_latest_citra_canary_with_camera_image/)
I've been trying to use it to scan QR codes in the game **Dragon Ball Heroes Ultimate Mission X** but it's not working.
All I get is a green screen and the behind the scenes window is showing these messages which keep scrolling infinitely [https://i.imgur.com/tqrwlvg.jpg]( https://i.imgur.com/tqrwlvg.jpg)
It works perfectly fine in Pokemon Sun so maybe the Dragon Ball game is using other functions of the camera or something?
Again sorry if this was not the right place to post this, hope this made some sense.

**Comment by Robz8** [comment]
Tried the build from the reddit link.
Can confirm AR card features work. Tested with Style Savvy: Trendsetters.
grace and renee hi-res

Style Savvy: Fashion Forward just shows a black screen, instead of the selected image for outer camera.

This continuously appears in the log for Trendsetters.
`Service.CAM <Warning> core/hle/service/cam/cam.cpp:GetTransferBytes:469: (STUBBED)called, port_select=1`
and this alongside the above, for Fashion Forward.
`Service.CAM <Warning> core/hle/service/cam/cam.cpp:GetTransferBytes:469: (STUBBED)called, port_select=2`

I'm hoping the other PR with camera integration will solve the black screen issue, in case if this PR doesn't get picked up. :P

**Comment by jroweboy** [comment]
@codename5 Thanks a bunch for this contribution! It seems that someone else has taken over this feature, and so I'll be closing this pull request in favor of the new one. https://github.com/citra-emu/citra/pull/3566 

I know you put good work into this, and I hope this doesn't dissuade you from wanting to contribute again in the future ❤️
